MINOR: Add JmxTool wrapper scripts and redirection (KIP-906)

fvaleri opened the pull request.

fvaleri committed:
Minor: Add JmxTool note to 3.5.0 notable changes

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

dajac commented on docs/upgrade.html:
```diff
@@ -19,6 +19,16 @@
+    <h5><a id="upgrade_350_notable" href="#upgrade_350_notable">Notable changes in 3.5.0</a></h5>
+    <ul>
+        <li>The JmxTool has been migrated to the tools module.
+            As there is no wrapper script, users have to update their commands from "kafka-run-class.sh kafka.tools.JmxTool" to "kafka-run-class.sh org.apache.kafka.tools.JmxTool".
```
It is a tad annoying that we are not backward compatible here. We usually never break compatibility without a deprecation period. Would it be imaginale to keep the old class name and to just call the new one? We could add a warning in the mean time and we could remove it in the next major release. This is just a thought.

fvaleri replied:
Hi @dajac, thanks for raising this. I agree, even if this is not one of the most popular tool and there has been some discussion about replacement, I've recently seen some people using it for debugging purpose and maybe for automation.

I updated the relevant note with the deprecation plan you suggests and changed `kafka-run-class` scripts so that we can print a warning and redirect to the new implementation class. There are other tools with no wrapper script and we can simply add an if for each of them. Let me know what you think of this approach.

@mimaison WDYT?

dajac replied:
Handling it in `kafka-run-class` does not seem to be the right way. Have you considered keeping the old class and using it to log a warning and to redirect?

fvaleri replied:
Yes I have and it's easier, but that would require adding tools dependency to core module and I have the impression that we want to avoid this.

mimaison replied:
For the sake of considering all ideas, if we think renaming these classes is problematic, another option would be to have Java classes in the tools module that match the old Scala classes name and package. This should allow existing commands to keep working while not requiring core to depend on tools.

I wonder if we should add scripts for all tools. That would separate the actual classes (which are not part of the public APIs) from the scripts which effectively are part of Kafka's public APIs.

dajac replied:
> For the sake of considering all ideas, if we think renaming these classes is problematic, another option would be to have Java classes in the tools module that match the old Scala classes name and package. This should allow existing commands to keep working while not requiring core to depend on tools.

I was thinking about the same. Then, we could add a warning in that class and remove it in the next major release.

fvaleri replied:
> For the sake of considering all ideas, if we think renaming these classes is problematic, another option would be to have Java classes in the tools module that match the old Scala classes name and package. 

I also tried this. The problem with this idea is that checkstyle is only able to handle a single root package per module, so you end up with the following error.

```sh
[ant:checkstyle] [ERROR] /home/fvaleri/Documents/kafka/tools/src/main/java/kafka/tools/JmxTool.java:17:1: Import control file does not handle this package. [ImportControl]
```

Possible alternatives: 

- create `:tools:deprecated` sub module containing `kafka.tools` root package with redirect classes (this works fine)
- use reflection inside core module to redirect without depending on tools module (have to try this one)

clolov commented on docs/upgrade.html:
```diff
@@ -19,6 +19,16 @@
+
+    <h5><a id="upgrade_350_notable" href="#upgrade_350_notable">Notable changes in 3.5.0</a></h5>
+    <ul>
+        <li>The JmxTool has been migrated to the tools module.
```
For curiosity, do you envision for us to add a similar `<li>` here for each tool as we move them across?

fvaleri replied:
No, we can simply add tool names that do not have a wrapper script to this list item (e.g. JmxTool and StateChangeLogMerger).

fvaleri commented:
@dajac @mimaison pushed the reflection based solution. Let me know if this is fine. Thanks.

ijuma commented:
Was a KIP submitted for this? Generally, if something ends up in the release notes due to user impact, it requires a KIP.

dajac commented:
I am not aware of any KIP regarding the migration of the tools from core module to tools module. Having a small one for those problematic tools is not a bad idea.

For the context, package name of the tools has been changed during the migration. For tools with a script, the change is transparent. For tools without a script, the change will break things. Hence my question in this PR.

ijuma commented:
@dajac well, that depends on whether it's considered public api. My position is that the only API is the script. If there is no script, it's not public API. How would users even know about the tools if they're not documented?

dajac commented:
Yeah, it is definitely a gray area. I don’t think that those tools are documented in the Kafka documentation. Somehow, JmxTool is known and you can find references to it on Stackoverflow and in articles online.

I don’t have a strong opinion on this but thoughts that it is worth having a discussion.

dajac commented:
By the way, there is another case here: https://github.com/apache/kafka/pull/13214.

Hangleton commented:
Thanks David for your suggestion about a KIP, this looks appropriate indeed. And, this can be an opportunity to add documentations for these SPIs, too.

fvaleri commented:
@ijuma @dajac I'm working on this KIP. Thanks for the feedback.

Hangleton commented:
@fvaleri Is the KIP going to include the interface `MessageReader`? Happy to write one for it otherwise.

fvaleri commented:
> `MessageReader`

Not specifically, I was thinking more about proposing a series of general guidelines to address the common compatibility issues we have seen while kicking off this migration activity.

fvaleri commented:
@ijuma @dajac 

https://cwiki.apache.org/confluence/display/KAFKA/KIP-906%3A+Tools+migration+guidelines

fvaleri committed:
Remove blank line

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri commented:
@dajac @ijuma can you please review the above small KIP? 

The vote thread is already opened.
Thanks.

fvaleri commented:
Hi @mimaison, now that KIP-906 is accepted I would say this is ready for the final review. Thanks.

Hangleton reviewed (approved):
Thanks for the PR.

fvaleri committed:
Add scripts

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri committed:
Use the new script in system tests

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

fvaleri commented:
@mimaison I also added the missing wrapper scripts for this tool as specified in KIP-906 and changed the system tests to use it. With this there is no more reference to the generic kafka-run-class script in STs.

System test run after the change:

```
TC_PATHS="tests/kafkatest/tests/core/throttling_test.py::ThrottlingTest.test_throttled_reassignment" \
  _DUCKTAPE_OPTIONS='--parameters '\''{"bounce_brokers":"false"}'\' \
    bash tests/docker/run_tests.sh
docker exec ducker01 bash -c "cd /opt/kafka-dev && ducktape --cluster-file /opt/kafka-dev/tests/docker/build/cluster.json  ./tests/kafkatest/tests/core/throttling_test.py::ThrottlingTest.test_throttled_reassignment --parameters '{"bounce_brokers":"false"}'"
/usr/local/lib/python3.9/dist-packages/paramiko/transport.py:236: CryptographyDeprecationWarning: Blowfish has been deprecated
  "class": algorithms.Blowfish,
[INFO:2023-04-12 06:15:49,034]: starting test run with session id 2023-04-12--001...
[INFO:2023-04-12 06:15:49,034]: running 1 tests...
[INFO:2023-04-12 06:15:49,035]: Triggering test 1 of 1...
[INFO:2023-04-12 06:15:49,042]: RunnerClient: Loading test {'directory': '/opt/kafka-dev/tests/kafkatest/tests/core', 'file_name': 'throttling_test.py', 'cls_name': 'ThrottlingTest', 'method_name': 'test_throttled_reassignment', 'injected_args': {'bounce_brokers': 'false'}}
[INFO:2023-04-12 06:15:49,043]: RunnerClient: kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false: on run 1/1
[INFO:2023-04-12 06:15:49,045]: RunnerClient: kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false: Setting up...
[INFO:2023-04-12 06:15:52,005]: RunnerClient: kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false: Running...
[INFO:2023-04-12 06:20:30,334]: RunnerClient: kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false: Tearing down...
[WARNING - 2023-04-12 06:20:30,669 - service_registry - stop_all - lineno:53]: Error stopping service <ProducerPerformanceService-0-139894120584000: num_nodes: 1, nodes: ['ducker09']>: 1
[INFO:2023-04-12 06:21:29,909]: RunnerClient: kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false: PASS
[INFO:2023-04-12 06:21:29,909]: RunnerClient: kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false: Data: None
================================================================================
SESSION REPORT (ALL TESTS)
ducktape version: 0.11.3
session_id:       2023-04-12--001
run time:         5 minutes 40.891 seconds
tests run:        1
passed:           1
flaky:            0
failed:           0
ignored:          0
================================================================================
test_id:    kafkatest.tests.core.throttling_test.ThrottlingTest.test_throttled_reassignment.bounce_brokers=false
status:     PASS
run time:   5 minutes 40.865 seconds
--------------------------------------------------------------------------------
```

mimaison commented:
Technically KIP-906 was voted after KIP freeze but since some tools have already been migrated I think we should merge PRs that add the redirection and the warning message to avoid breaking things.

mimaison commented:
@dajac Any objections?

dajac commented:
@mimaison No objections. I do agree that we should avoid breaking things.

fvaleri committed:
Revert ST changes

Signed-off-by: Federico Valeri <fedevaleri@gmail.com>

mimaison reviewed (approved):
LGTM

mimaison commented:
Since the initial JmxTool changes are in 3.5, I've backported this to 3.5 too.
